### PR TITLE
util: add AmazonMetadata KMS provider

### PR DESCRIPTION
**Pull request by nixpanic**
The new Amazon Metadata KMS provider uses a CMK stored in AWS KMS to
encrypt/decrypt the DEK which is stored in the volume metadata.

Updates: #1921
~Depends-on: #1927~

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

### CODE REVIEW

**Comment by nixpanic** [comment]
> /retest all

Not sure what happened, all jobs were aborted somehow.

**Comment by humblec** [comment]
@nixpanic `... CMK stored in AWS KMS to`.. by CMK you meant custom master key, Isnt it ?

**Comment by nixpanic** [comment]
> @nixpanic `... CMK stored in AWS KMS to`.. by CMK you meant custom master key, Isnt it ?

Amazon calls is [Customer Master Key](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#master_keys).

**Comment by nixpanic** [comment]
A test image build from this PR is available:
```
podman pull quay.io/nixpanic/cephcsi:testing_kms_AWSMetadata
```

File: examples/kms/vault/aws-credentials.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# Namespace where Ceph-CSI is deployed. The contents of this Secret will be
+# used to connect to the Amazon KMS.
+apiVersion: v1
+kind: Secret
+metadata:
+  name: ceph-csi-aws-credentials
+stringData:
+  AWS_ACCESS_KEY_ID: "arn:aws:kms:us-west-2:111122223333:key/1234abcd-12ab-34cd-56ef-1234567890ab"
+  AWS_SECRET_ACCESS_KEY: "JalrXUtnFEMI/K7MDENG/bPxRfiCYzEXAMPLEKEY"
+  AWS_SESSION_TOKEN: ""
```
Comment (humblec) [inline examples/kms/vault/aws-credentials.yaml:12]: @nixpanic can we document about these and also mention which are mandatory/required?

File: examples/kms/vault/csi-kms-connection-details.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+#
+# csi-kms-connection-details is an alternative option to configure KMS
+# providers for encrypted volume support.
+# This ConfigMap can be located in the Kubernetes Namespace where Ceph-CSI is
+# deployed. In case the ceph-csi-encryption-kms-config which provides a
```
Comment (humblec) [inline examples/kms/vault/csi-kms-connection-details.yaml:5]: @nixpanic why we have fallback configmap ? In which scenarios it helps?

File: examples/kms/vault/csi-kms-connection-details.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      "KMS_PROVIDER": "vaulttokens",
+      "VAULT_ADDR": "http://vault.default.svc.cluster.local:8200",
+      "VAULT_BACKEND_PATH": "secret",
+      "VAULT_SKIP_VERIFY": "true"
+    }
+  secrets-metadata-test: |-
+    {
+      "encryptionKMSType": "metadata"
+    }
+  aws-metadata-test: |-
```
Comment (humblec) [inline examples/kms/vault/csi-kms-connection-details.yaml:38]: @nixpanic can we have some documentation about these sections?

File: internal/util/aws_metadata.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+	}
+
+	err = setConfigString(&kms.secretName, args.Config, awsSecretKey)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	} else if errors.Is(err, errConfigOptionMissing) {
+		kms.secretName = awsMetadataDefaultSecretsName
+	}
+
+	// awsRegionKey is optional
```
Comment (humblec) [inline internal/util/aws_metadata.go]: @nixpanic I was an impression that region is required. Just wanted to confirm is it an optional value ?

File: examples/kms/vault/csi-kms-connection-details.yaml (hunk 2 above)
Comment (nixpanic) [reply examples/kms/vault/csi-kms-connection-details.yaml:5]: #1793 is still open and should describe the introduction of the ConfigMap. There was no example included with the implementation, making it difficult for users to understand how it works. This example now includes different KMS descriptions as an example.

File: examples/kms/vault/csi-kms-connection-details.yaml (hunk 3 above)
Comment (nixpanic) [reply examples/kms/vault/csi-kms-connection-details.yaml:38]: Yes, I would like to include that. Do you know which document describes the existing KMS providers? I can extend that.

File: internal/util/aws_metadata.go (hunk 4 above)
Comment (nixpanic) [reply internal/util/aws_metadata.go]: I thought I read in the AWS API docs it was optional?

It seems that it is optional for some calls, but probably not for the KMS API. I'll make it a mandatory option.

File: examples/kms/vault/csi-kms-connection-details.yaml (hunk 3 above)
Comment (nixpanic) [reply examples/kms/vault/csi-kms-connection-details.yaml:38]: Added this to `docs/deploy-rbd.md` now.

File: examples/kms/vault/aws-credentials.yaml (hunk 1 above)
Comment (nixpanic) [reply examples/kms/vault/aws-credentials.yaml:12]: I have added this to `docs/deploy-rbd.md`.

**Comment by nixpanic** [comment]
@nbalacha documentation and examples are available now too. Please have a look and provide feedback once you were able to test things out. Thanks!

File: examples/kms/vault/csi-kms-connection-details.yaml (hunk 3 above)
Comment (humblec) [reply examples/kms/vault/csi-kms-connection-details.yaml:38]: @nixpanic do we need to update https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/encrypted-pvc.md too?

File: docs/deploy-rbd.md [hunk 5]
```diff
@@ -280,6 +280,34 @@ Configure a role(s) for service accounts used for ceph-csi:
 * nodeplugin service account (`rbd-csi-nodeplugin`) requires **create** and
   **read** permissions to save new keys and retrieve existing
 
+#### Configuring Amazon KMS
+
+Amazon KMS can be used to encrypt and decrypt the passphrases that are used for
+encrypted RBD images. When a volume is created, a passphrase will be generated,
+which will be encrypted by the KMS and stored in the volumes metadata. Opon
```
Comment (humblec) [inline docs/deploy-rbd.md]: typo `Opon`

File: docs/deploy-rbd.md [hunk 6]
```diff
@@ -280,6 +280,34 @@ Configure a role(s) for service accounts used for ceph-csi:
...
+attaching the volume to a Pod, the worker node requests the KMS to decrypt the
+passphrase, after which it can be used to open the device with `cryptsetup` and
+provide access to it for the Pod.
+
+There are a few settings that need to be included in the [KMS configuration
+file](../examples/kms/vault/kms-config.yaml):
+
+1. `KMS_PROVIDER`: should be set to `aws-metadata`.
```
Comment (humblec) [inline docs/deploy-rbd.md:295]: do we need `aws-metadata` or just `aws` enough?

File: docs/deploy-rbd.md [hunk 7]
```diff
@@ -280,6 +280,34 @@ Configure a role(s) for service accounts used for ceph-csi:
...
+1. `AWS_REGION`: the region where the AWS KMS service is available.
+
+The [Secret with credentials](../examples/kms/vault/aws-credentials.yaml) for
+the AWS KMS is expected to contain:
+
+1. `AWS_ACCESS_KEY_ID`: ID of the key to use for encrypting/decrypting
+1. `AWS_SECRET_ACCESS_KEY`: secret for the key to use
+1. `AWS_SESSION_TOKEN`: *(optional)* session token, usually empty
+
+This Secret is expected to be created by the administrator who deployed
```
Comment (humblec) [inline docs/deploy-rbd.md:310]: can this be in any Namespace?

File: internal/util/aws_metadata.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+	awsSession "github.com/aws/aws-sdk-go/aws/session"
+	awsKMS "github.com/aws/aws-sdk-go/service/kms"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	kmsTypeAWSMetadata = "aws-metadata"
+
+	// awsMetadataDefaultSecretsName is the name of the Kubernetes Secret
+	// that contains the Vault Token for the tenant.  This Secret is
```
Comment (humblec) [inline internal/util/aws_metadata.go]: Is it called `vault token` even in case of `aws` @nixpanic ?

File: docs/deploy-rbd.md (hunk 5 above)
Comment (nixpanic) [reply docs/deploy-rbd.md]: Ah, will fix!

File: docs/deploy-rbd.md (hunk 6 above)
Comment (nixpanic) [reply docs/deploy-rbd.md:295]: At one point we might want to add a KMS implementation that does not store the DEK in the metadata of the volume. Just `aws` would be more suitable for that.

File: docs/deploy-rbd.md (hunk 7 above)
Comment (nixpanic) [reply docs/deploy-rbd.md:310]: No, it needs to be in the namesspace where Ceph-CSI is deployed.

File: internal/util/aws_metadata.go (hunk 8 above)
Comment (nixpanic) [reply internal/util/aws_metadata.go]: This has been corrected now too.

File: docs/deploy-rbd.md (hunk 6 above)
Comment (humblec) [reply docs/deploy-rbd.md:295]: Sounds good!

File: docs/deploy-rbd.md (hunk 7 above)
Comment (humblec) [reply docs/deploy-rbd.md:310]: @nixpanic do we need to mention?

File: internal/util/aws_metadata.go (hunk 8 above)
Comment (humblec) [reply internal/util/aws_metadata.go]: Thanks.

File: docs/deploy-rbd.md (hunk 7 above)
Comment (nixpanic) [reply docs/deploy-rbd.md:310]: No, I do not think so. If placing it in a different namespace was an option, *that* would need mentioning. Normally all objects for a deployment are located in the same (here ceph-csi) namespace.

File: internal/util/aws_metadata.go [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+// EncryptDEK uses the Amazon KMS and the configured CMK to encrypt the DEK.
+func (kms *AWSMetadataKMS) EncryptDEK(volumeID, plainDEK string) (string, error) {
+	svc, err := kms.getService()
+	if err != nil {
+		return "", fmt.Errorf("could not get KMS service: %w", err)
+	}
+
+	result, err := svc.Encrypt(&awsKMS.EncryptInput{
+		KeyId:     aws.String(kms.accessKey),
```
Comment (nbalacha) [inline internal/util/aws_metadata.go]: This needs to be an ARN for the CMK, which is different from the aws_access_key_id.
There needs to be an additional parameter for the CMK ARN in the secret.

**Review by nbalacha — CHANGES_REQUESTED** [review]

File: docs/deploy-rbd.md (hunk 7 above)
Comment (humblec) [reply docs/deploy-rbd.md:310]: sure.

File: internal/util/aws_metadata.go (hunk 9 above)
Comment (humblec) [reply internal/util/aws_metadata.go]: @nbalacha so do we need both `cmk` key and `access key` for the encryption or just `cmk` key is enough ?

Comment (nbalacha) [reply internal/util/aws_metadata.go]: We need both. The cmk is the actual key that will be used to perform the encryption/decryption. The access_key is used as part of the credentials to authenticate the user.

Comment (nbalacha) [reply internal/util/aws_metadata.go]: @humblec , example here: https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/kms-example-encrypt-data.html

Comment (humblec) [reply internal/util/aws_metadata.go]: Thanks @nbalacha , it helps :+1:

Comment (nbalacha) [reply internal/util/aws_metadata.go]: @nixpanic, How do I set up encryption for PVs? From the OCP console it looks like this needs to be set up in the storage class. 
1. Can there be a separate CMK per storageclass ?
2. Will the CMK be an entry in the secret or in something related to the storageclass?

Comment (nixpanic) [reply internal/util/aws_metadata.go]: Thank @nbalacha! Is there a standard environment variable that is used for the CMK-ARN? Or, can you recommend a variable that you would like to see used?

Comment (nbalacha) [reply internal/util/aws_metadata.go]: AWS_CMK_ARN or AWS_KEY_ID are possible options.

File: go.mod [hunk 10]
```diff
@@ -3,6 +3,7 @@ module github.com/ceph/ceph-csi
 go 1.13
 
 require (
+	github.com/aws/aws-sdk-go v1.35.24
```
Comment (Madhu-1) [inline go.mod]: the latest available release is 1.38.8 can we use that?

Comment (nixpanic) [reply go.mod]: I don't know, this was picked automatically by `go mod`?

File: internal/util/aws_metadata.go (hunk 9 above)
Comment (nixpanic) [reply internal/util/aws_metadata.go]: will add a `AWS_CMK_ARN` option

File: go.mod (hunk 10 above)
Comment (nixpanic) [reply go.mod]: will update to take the most current version

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.19

Deplying Rook failed with `etcdserver: request timed out` #1636

**Review by Madhu-1 — DISMISSED** [review]

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.19

The minikube node decided to go on holidays, or at least took some time off:

    Waiting up to 3m0s for all (but 0) nodes to be ready

**Review by nixpanic — COMMENTED** [review]
```go
err := setConfigString(&kms.namespace, args.Config, args.Namespace)
```

The namespace simply needs to get assigned, not fetched from a configuration file/map. Nithya reported the following error:
```
Warning  ProvisioningFailed    47s (x11 over 9m19s)  openshift-storage.rbd.csi.ceph.com_csi-rbdplugin-provisioner-6fb48ddd4d-ks8hb_bdaeb2d7-b0d0-4625-ab98-4c2accc7132b  failed to provision volume with StorageClass "kms-test-sc": rpc error: code = InvalidArgument desc = invalid encryption kms configuration: configuration option not set: openshift-storage
```

**Review by nbalacha — COMMENTED** [review]
I have tested this on an OCP cluster on AWS.
1. Created a CMK in my AWS account
2. Created a configmap using the specified format
3. Created a secret with the AWS key, secretkey and CMK details
4. Created a storageclass with encryption enabled and created a PVC using the sc
5. Created a Pod to use the PVC.

Result:
The PVC is created and bound successfully
The pod is able to mount the PVC
I am able to create files in the volume

**Comment by humblec** [comment]
> I have tested this on an OCP cluster on AWS.
> 
>     1. Created a CMK in my AWS account
> 
>     2. Created a configmap using the specified format
> 
>     3. Created a secret with the AWS key, secretkey and CMK details
> 
>     4. Created a storageclass with encryption enabled and created a PVC using the sc
> 
>     5. Created a Pod to use the PVC.
> 
> 
> Result:
> The PVC is created and bound successfully
> The pod is able to mount the PVC
> I am able to create files in the volume

Thanks @nbalacha . 

If can be performed, few more checks I  would like to propose with this change:

1) Deletion of POD/App and making sure the pod reattach to another node.
2) Deletion of PVC , then verification on KMS side( if it make sense).

3) Also if possible we can quickly verify snapshot/clone request on this PVC and how it behaves ( based on your time)

**Comment by nbalacha** [comment]
> If can be performed, few more checks I would like to propose with this change:
> 
> 1. Deletion of POD/App and making sure the pod reattach to another node.

I deleted the pod and recreated it setting the node affinity to another node. It came up without issues.

> 2. Deletion of PVC , then verification on KMS side( if it make sense).

Can you elaborate on this? What kind of verification did you have in mind?

> 3. Also if possible we can quickly verify snapshot/clone request on this PVC and how it behaves ( based on your time)

While creating an SC, the UI states that "Encrypted PVs cannot be cloned, expanded or used to create snapshots" . That is for Vault but I am assuming the same would hold good here as well.


I did see an issue where the "Create Storage Class" page in the UI goes blank if I select encryption if the following Config Map exists in the namespace. I intend to follow up with the UI team on that.

apiVersion: v1
kind: ConfigMap
metadata:
  name: csi-kms-connection-details
data:
  aws-metadata-test: |-
    {   
      "KMS_PROVIDER": "aws-metadata",
      "KMS_SECRET_NAME": "ceph-csi-aws-credentials",
      "AWS_REGION": "us-east-2"
    } 

@humblec

**Comment by nixpanic** [comment]
>     3\. Also if possible we can quickly verify snapshot/clone request on this PVC and how it behaves ( based on your time)

This is not expected to work yet, I am still working on #1470.

**Comment by nixpanic** [comment]
Hi @humblec @Madhu-1, this is ready for review.

@nbalacha successfully tested this manually. As it requires an Amazon KMS, we can not run tests in our CI :disappointed:

**Review by Madhu-1 — APPROVED** [review]

**Comment by humblec** [comment]
> > If can be performed, few more checks I would like to propose with this change:
> > 
> > 1. Deletion of POD/App and making sure the pod reattach to another node.
> 
> I deleted the pod and recreated it setting the node affinity to another node. It came up without issues.

Thanks for verifying this scenario! 

> 
> > 1. Deletion of PVC , then verification on KMS side( if it make sense).
> 
> Can you elaborate on this? What kind of verification did you have in mind?

I was thinking about the key verification at KMS engine side, but looks like we are good with the current verification.

> 
> > 1. Also if possible we can quickly verify snapshot/clone request on this PVC and how it behaves ( based on your time)
> 
> While creating an SC, the UI states that "Encrypted PVs cannot be cloned, expanded or used to create snapshots" . That is for Vault but I am assuming the same would hold good here as well.
> 
> I did see an issue where the "Create Storage Class" page in the UI goes blank if I select encryption if the following Config Map exists in the namespace. I intend to follow up with the UI team on that.
> 
> apiVersion: v1
> kind: ConfigMap
> metadata:
> name: csi-kms-connection-details
> data:
> aws-metadata-test: |-
> {
> "KMS_PROVIDER": "aws-metadata",
> "KMS_SECRET_NAME": "ceph-csi-aws-credentials",
> "AWS_REGION": "us-east-2"
> }

This could be a different issue I believe. Thanks for reporting it to the UI team as well.

**Comment by humblec** [comment]
> Hi @humblec @Madhu-1, this is ready for review.
> 
> @nbalacha successfully tested this manually. As it requires an Amazon KMS, we can not run tests in our CI disappointed

Yeah, for now lets move ahead with the manual verification done already here.

**Review by humblec — APPROVED** [review]

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e/k8s-1.19

The minikube VM became unresponsive?
```
Apr  5 08:55:08.733: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
```

**Comment by nixpanic** [comment]
> /retest ci/centos/upgrade-tests-rbd

Failed to deploy Rook #1636

**Pull request merged**